### PR TITLE
Feature/liayoo/ainft v1

### PR DESCRIPTION
**Pull request by liayoo**
- Added AINFTv1.sol and AINFTv1Mock contracts
- Added tests for AINFTv1Mock.sol
- To be published to the npm registry and installed as a package (like https://www.npmjs.com/package/@openzeppelin/contracts)

### CODE REVIEW

File: contracts/ainft/AINFTv1.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+   * @param offset_ index offset to start enumerating within the ownedTokens list of the owner
+   * @param limit_ max number of IDs to fetch
+   */
+  function tokensOf(
+    address owner_,
+    uint256 offset_,
+    uint256 limit_
+  ) public view returns (uint256[] memory) {
+    uint256 balance = ERC721.balanceOf(owner_);
+    require(limit_ <= TOKEN_FETCH_LIMIT, "AINFTv1: limit too large");
```
Comment (platfowner) [inline contracts/ainft/AINFTv1.sol]: minor) could the error message include TOKEN_FETCH_LIMIT somehow? (let me know if it's impossible in solidity)

File: contracts/ainft/AINFTv1.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    require(offset_ < balance, "AINFTv1: invalid offset");
```
Comment (platfowner) [inline contracts/ainft/AINFTv1.sol]: minor) ditto for balance?

File: contracts/ainft/AINFTv1.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    for (uint256 i = 0; i < numToReturn; i++) {
+      ownedTokens[i] = tokenOfOwnerByIndex(owner_, offset_ + i);
+    }
+    return ownedTokens;
+  }
+
+  // ============= TX
+
+  function setBaseURI(string calldata baseURI_) public onlyRole(OWNER_ROLE) {
+    require(bytes(baseURI_).length > 0, "AINFTv1: invalid value");
```
Comment (platfowner) [inline contracts/ainft/AINFTv1.sol]: minor) what about something like "invalid baseURI_"?

File: contracts/ainft/AINFTv1.sol [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    baseURI = baseURI_;
+  }
+
+  function setMaxTokenId(uint256 maxTokenId_) public onlyRole(OWNER_ROLE) {
+    require(nextTokenId - 1 <= maxTokenId_, "AINFTv1: invalid value");
```
Comment (platfowner) [inline contracts/ainft/AINFTv1.sol]: minor) what about something like "invalid maxTokenId_"?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: contracts/ainft/AINFTv1.sol (hunk 2 above)
Comment (liayoo) [reply contracts/ainft/AINFTv1.sol:68]: done!

File: contracts/ainft/AINFTv1.sol (hunk 3 above)
Comment (liayoo) [reply contracts/ainft/AINFTv1.sol]: done!

File: contracts/ainft/AINFTv1.sol (hunk 4 above)
Comment (liayoo) [reply contracts/ainft/AINFTv1.sol]: done!

File: contracts/ainft/AINFTv1.sol (hunk 1 above)
Comment (liayoo) [reply contracts/ainft/AINFTv1.sol]: I've defined a custom error for this! See https://github.com/ainft-team/ainft-contract-template/pull/1/files#diff-96b23af4415b09de07a9952eaf6c4a5de05f3b32c409a98fe68a66a77603d8c4R170 for reference.

File: contracts/ainft/AINFTv1.sol (hunk 2 above)
Comment (liayoo) [reply contracts/ainft/AINFTv1.sol]: ditto

File: contracts/ainft/AINFTv1.sol (hunk 1 above)
Comment (platfowner) [reply contracts/ainft/AINFTv1.sol]: 👍

File: contracts/ainft/AINFTv1.sol (hunk 2 above)
Comment (platfowner) [reply contracts/ainft/AINFTv1.sol]: 👍

**Comment by liayoo** [comment]
Thanks for the review!

**Pull request merged**